src/v/cloud_storage/remote_partition.cc

```diff
@@ -69,6 +71,134 @@ get_kafka_max_offset(const partition_manifest::segment_meta& m) {
+                mit = _manifest.segment_containing(ko);
+            }
+            if (mit == _manifest.end()) {
+                // Segment that matches exactly can't be found
+                // in the manifest. In this case we want to start
+                // scanning from the begining of the partition if
+                // some conditions are met.
+                auto so = _manifest.get_start_kafka_offset().value_or(
+                  kafka::offset::min());
+                if (config.start_offset < so && config.max_offset > so) {
+                    mit = _manifest.begin();
+                }
+            }
+        } else {
+            mit = _manifest.segment_containing(hint);
+            while (mit != _manifest.end()) {
+                // The segment 'mit' points to might not have any
+                // data batches. In this case we need to move iterator forward.
+                // The check can only be done if we have 'delta_offset_end'.
+                if (mit->second.delta_offset_end == model::offset_delta{}) {
+                    break;
+                }
+                auto b = mit->second.base_offset - mit->second.delta_offset;
+                auto m = mit->second.committed_offset
+                         - mit->second.delta_offset_end;
+                if (b != m) {
+                    break;
+                }
+                mit++;
+            }
+        }
+        if (mit == _manifest.end()) {
+            // No such segment
+            return borrow_result_t{};
+        }
+        auto iter = _segments.find(mit->first);
+        if (iter != _segments.end()) {
+            if (
+              iter->second->segment->get_max_rp_offset()
+              != mit->second.committed_offset) {
+                offload_segment(iter->first);
+                iter = _segments.end();
+            }
+        }
+        if (iter == _segments.end()) {
+            iter = materialize_segment(mit->second);
+        }
+        auto next_it = std::next(mit);
+        while (next_it != _manifest.end()) {
+            // Normally, the segments in the manifest do not overlap.
+            // But in some cases we may see them overlapping, for instance
+            // if they were produced by older version of redpanda.
+            // In this case we want to skip segment if its offset range
+            // lies withing the offset range of the current segment.
+            if (
+              mit->second.committed_offset < next_it->second.committed_offset) {
+                break;
+            }
+            next_it++;
+        }
+        model::offset next_offset = next_it == _manifest.end() ? model::offset{}
+                                                               : next_it->first;
+        return borrow_result_t{
+          .reader = iter->second->borrow_reader(config, _ctxlog, _probe),
+          .next_segment_offset = next_offset};
+    };
+    return kafka_offset_lookup(config);
```

Comment:
hmm, not sure what are the benefits of declaring a lambda and immediately calling it?

Reply:
I had two lambdas  that were called conditionally. Then I removed one of them. Will fix.